File: Source/frontend/python_backend/app/routers/resources.py
"""Resource endpoints."""
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import text
from typing import Optional, List

from ..database import get_db
from ..schemas import ResourcesResponse
from ..utils import parse_point

router = APIRouter(prefix="/api", tags=["Resources"])


@router.get("/resources", response_model=ResourcesResponse)
async def get_resources(
    type: Optional[str] = Query(None, description="Filter by resource type"),
    status: Optional[str] = Query(None, description="Filter by status"),
    db: AsyncSession = Depends(get_db)
):
    """Get resources with optional filters."""
    try:
        params = {}
        clauses = []
        
        if type:
            clauses.append("LOWER(type) = LOWER(:type)")
            params["type"] = type
        
        if status:
            clauses.append("LOWER(status) = LOWER(:status)")
            params["status"] = status
        
        where_clause = ""
        if clauses:
            where_clause = "WHERE " + " AND ".join(clauses)
        
        query = text(f"""
            SELECT
                code,
                name,
                type,
                status,
                capacity,
                capabilities,
                contact_info,
                ST_AsGeoJSON(location, 6) as location
            FROM resources
            {where_clause}
            ORDER BY type, name
        """)
        
        result = await db.execute(query, params)
        rows = result.fetchall()
        
        payload = {
            "depots": [],
            "equipment": [],
            "crews": []
        }
        
        for row in rows:
            location = parse_point(row.location)
            capabilities = row.capabilities or {}
            
            if row.type == 'depot':
                payload["depots"].append({
                    "id": row.code,
                    "name": row.name,
                    "lat": location['lat'],
                    "lng": location['lng'],
                })
            elif row.type == 'equipment':
                payload["equipment"].append({
                    "id": row.code,
                    "type": capabilities.get('type', row.name),
                    "subtype": capabilities.get('subtype'),
                    "capacity_lps": capabilities.get('capacity_lps'),
                    "units": capabilities.get('units'),
                    "depot": capabilities.get('depot'),
                    "status": row.status,
                })
            elif row.type == 'crew':
                payload["crews"].append({
                    "id": row.code,
                    "name": row.name,
                    "skills": capabilities.get('skills', []),
                    "depot": capabilities.get('depot'),
                    "status": row.status,
                    "lat": location['lat'],
                    "lng": location['lng'],
                })
        
        return payload
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to fetch resources: {str(e)}")

File: Source/frontend/backend/routes/resources.js
import { Router } from 'express';
import { parsePoint } from '../utils/helpers.js';

const router = Router();

/**
 * @swagger
 * /api/resources:
 *   get:
 *     tags: [Resources]
 *     summary: Get resources with optional filters
 *     description: Retrieve depots, equipment, and crews with optional filtering
 *     parameters:
 *       - in: query
 *         name: type
 *         schema:
 *           type: string
 *           enum: [depot, equipment, crew]
 *         description: Resource type filter
 *       - in: query
 *         name: status
 *         schema:
 *           type: string
 *         description: Resource status filter
 *     responses:
 *       200:
 *         description: Successful response
 *         content:
 *           application/json:
 *             schema:
 *               type: object
 *               properties:
 *                 depots:
 *                   type: array
 *                   items:
 *                     type: object
 *                     properties:
 *                       id:
 *                         type: string
 *                       name:
 *                         type: string
 *                       lat:
 *                         type: number
 *                       lng:
 *                         type: number
 *                 equipment:
 *                   type: array
 *                   items:
 *                     type: object
 *                     properties:
 *                       id:
 *                         type: string
 *                       type:
 *                         type: string
 *                       subtype:
 *                         type: string
 *                       capacity_lps:
 *                         type: number
 *                       units:
 *                         type: number
 *                       depot:
 *                         type: string
 *                       status:
 *                         type: string
 *                 crews:
 *                   type: array
 *                   items:
 *                     type: object
 *                     properties:
 *                       id:
 *                         type: string
 *                       name:
 *                         type: string
 *                       skills:
 *                         type: array
 *                         items:
 *                           type: string
 *                       depot:
 *                         type: string
 *                       status:
 *                         type: string
 *                       lat:
 *                         type: number
 *                       lng:
 *                         type: number
 */
router.get('/', async (req, res) => {
  try {
    const { type, status } = req.query;

    let query = `
      SELECT
        code,
        name,
        type,
        status,
        capacity,
        capabilities,
        contact_info,
        ST_AsGeoJSON(location, 6) as location
      FROM resources
    `;

    const params = [];
    const clauses = [];
    let paramIndex = 1;

    if (type) {
      clauses.push(`LOWER(type) = LOWER($${paramIndex++})`);
      params.push(type);
    }

    if (status) {
      clauses.push(`LOWER(status) = LOWER($${paramIndex++})`);
      params.push(status);
    }

    if (clauses.length > 0) {
      query += ` WHERE ${clauses.join(' AND ')}`;
    }

    query += ' ORDER BY type, name';

    const result = await req.pool.query(query, params);

    const payload = {
      depots: [],
      equipment: [],
      crews: []
    };

    result.rows.forEach(row => {
      const location = parsePoint(row.location);
      const capabilities = row.capabilities || {};

      if (row.type === 'depot') {
        payload.depots.push({
          id: row.code,
          name: row.name,
          lat: location.lat,
          lng: location.lng,
        });
        return;
      }

      if (row.type === 'equipment') {
        payload.equipment.push({
          id: row.code,
          type: capabilities.type || row.name,
          subtype: capabilities.subtype,
          capacity_lps: capabilities.capacity_lps,
          units: capabilities.units,
          depot: capabilities.depot,
          status: row.status,
        });
        return;
      }

      if (row.type === 'crew') {
        payload.crews.push({
          id: row.code,
          name: row.name,
          skills: capabilities.skills || [],
          depot: capabilities.depot,
          status: row.status,
          lat: location.lat,
          lng: location.lng,
        });
      }
    });

    res.json(payload);
  } catch (error) {
    console.error('Database error:', error);
    res.status(500).json({
      error: 'Failed to fetch resources',
      details: process.env.NODE_ENV === 'development' ? error.message : undefined
    });
  }
});

export default router;